### PR TITLE
feat: ✨ annotate container issues in Kubernetes files and display them in problem view [ROAD-399]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
image
image
image
Annotation tooltip/message text has been approved by Andy.

### CODE REVIEW

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import snyk.container.ContainerIssuesForImage
+
+class ContainerYamlAnnotator : ExternalAnnotator<PsiFile, Unit>() {
+    private val logger = logger<ContainerYamlAnnotator>()
+
+    // overrides needed for the Annotator to invoke apply(). We don't do anything here
+    override fun collectInformation(file: PsiFile): PsiFile? = file
+    override fun doAnnotate(collectedInfo: PsiFile?): Unit = Unit
+
+    fun getContainerIssuesForImages(psiFile: PsiFile): List<ContainerIssuesForImage> {
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:21]: internal ?

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            val severity = severity(forImage)
+
+            val workloadImage = forImage.workloadImages.first { it.psiFile == psiFile }
+            holder.newAnnotation(severity, annotationMessage(forImage))
+                .range(textRange(psiFile, workloadImage.lineNumber, forImage.imageName))
+                .create()
+        }
+    }
+
+    @TestOnly
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:50]: internal ? `@TestOnly` is not really true here as it used in apply()

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        return when {
+            severities[SEVERITY_CRITICAL]?.isNotEmpty() == true -> HighlightSeverity.ERROR
+            severities[SEVERITY_HIGH]?.isNotEmpty() == true -> HighlightSeverity.WARNING
+            severities[SEVERITY_MEDIUM]?.isNotEmpty() == true -> HighlightSeverity.WEAK_WARNING
+            severities[SEVERITY_LOW]?.isNotEmpty() == true -> HighlightSeverity.INFORMATION
+            else -> HighlightSeverity.INFORMATION
+        }
+    }
+
+    fun annotationMessage(image: ContainerIssuesForImage): String {
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:62]: internal ?

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        return buildString {
+            val remediationString = when {
+                image.baseImageRemediationInfo != null -> "Upgrade image to a newer version"
+                else -> ""
+            }
+            append("Snyk found $total $vulnerabilityString. $remediationString")
+        }
+    }
+
+    fun textRange(psiFile: PsiFile, line: Int, imageName: String): TextRange {
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:75]: internal ?

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            holder.newAnnotation(severity, annotationMessage(forImage))
+                .range(textRange(psiFile, workloadImage.lineNumber, forImage.imageName))
+                .create()
+        }
+    }
+
+    @TestOnly
+    fun severity(forImage: ContainerIssuesForImage): HighlightSeverity {
+        val severities = forImage.vulnerabilities.groupBy { it.severity }
+        return when {
```
Comment (ArtsiomCh) [inline src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:53]: minor: some kind of comparator could be usefull here (also could be reused for other products)

**Review by ArtsiomCh — APPROVED** [review]
LGTM
@bastiandoetsch just some minor comments could be addressed if found needed.

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt (hunk 3 above)
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt]: needed for testing - not possible :)

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt (hunk 4 above)
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:73]: needed for testing - not possible :)

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt (hunk 5 above)
Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:51]: Let's write that comparator once we need it the second time!

File: src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt (hunk 3 above)
Comment (ArtsiomCh) [reply src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt]: hmm.. test is in the same package, so `internal` should works for both class and test usage.

Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:60]: Tried it out - compile error

Comment (bastiandoetsch) [reply src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:60]: (and thought exactly the same - I changed it first, but didn't work)

Comment (ArtsiomCh) [reply src/main/kotlin/snyk/container/annotator/ContainerYamlAnnotator.kt:60]: ah, it's cause it in integTest that doesn't see as the same package for some reason. Ok. let's leave it public then

**Pull request merged**